### ISSUE TITLE
Performance issues with hash sets in 2.0 without j4ts

### ISSUE THREAD

**Issue by DavidRigglemanININ** [cincheo/jsweet#304]
Looking at the implementation in TS of hash sets without j4ts, it appears that the contains and remove operations are just using indexOf (which does a full scan of the list). This is an O(n) operation when the whole purpose of using sets is that these operations should be O(1). With the current implementation, the hash set implementation is no better than just using a list. 

For reference, we had to implement our own set implementation in TS and we the way we did it briefly summarizing was using `this.elements[key] = val` for the add, `Object.prototype.hasOwnProperty.call(this.elements, key);` for the contains, `delete this.elements[key]` for the delete, where this.elements is just a plain JS map: {}

**Comment by DavidRigglemanININ** [comment]
I should add to the above that our implementation of set is basically a wrapper around map where the value is a boolean always set to true. This actually closely matches up with the JVM implementation of HashSet: https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/HashSet.java

**Comment by renaudpawlak** [comment]
Actually, we have made the choice so far to avoid using JavaScript maps, just to be compatible with older browsers (we have made the same choice with Java maps actually). This strategy is of course sub-optimal. However, with JSweet v2, it is quite easy to create an adapter that uses the JS Map and leads to much better performance.

Actually I was planning to use such an adapter as an example of how to write a JSweet 2 adapter to tune the way JSweet generates code for your own context. So maybe you want to give it a try, or just wait for the example :)

**Comment by DavidRigglemanININ** [comment]
That doesn't seem true, though, at least your response about making the same choice for Java maps

```
HashMap<String, Boolean> set = new HashMap<String, Boolean>();
set.put("TEST", true);
boolean containsValue = set.containsKey("TEST1");
set.remove("TEST");
```
gets transpiled to TS as

```
let set : any = <any>({});
/* put */(set["TEST"] = true);
let containsValue : boolean = /* containsKey */set.hasOwnProperty("TEST1");
/* remove */delete set["TEST"];
```

**Comment by renaudpawlak** [comment]
Yes. That's because you are using a HashMap with keys that are strings. In that case, JSweet knows that it can use a plain JS Object to emulate the map (note that it is not a JS [Map](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Map), though, it is a JS Object). The optimisation is quite straightforward, here is the [test](https://github.com/cincheo/jsweet/blob/master/transpiler/src/main/java/org/jsweet/transpiler/extension/RemoveJavaDependenciesAdapter.java#L735) to know if we are having string keys or not.

We could have a similar optimisation for HashSet when it is containing strings only (easy indeed), but I thought that you were talking in the general case.

**Comment by DavidRigglemanININ** [comment]
Yeah, I guess I was thinking more about string HashSets and having that same optimization. It's definitely more tricky to make it work with non string keys. Our issues would be resolved if the same optimization for maps was made for sets

**Comment by renaudpawlak** [comment]
Definitely. A PR would be welcome on that one ;)

**Comment by DavidRigglemanININ** [comment]
So the change would go in RemoveJavaDependenciesAdapter.java, right? Currently it looks like the Set code is implemented via substituteMethodInvocationOnArray. So I'm thinking we'd want a new substituteMethodInvocationOnSet function that is only called if the key type was string.

**Comment by lgrignon** [comment]
Thanks for your contribution!

**Comment by lgrignon** [comment]
Not only on substituteMethodInvocationXX but also substituteNewClass ;)

**Comment by renaudpawlak** [comment]
Yes, ``RemoveJavaDependenciesAdapter.java``. You may want to create a new method or keep it in ``substituteMethodInvocationOnArray`` (I am not sure that it is the best naming).
Actually, the first thing to do would be to map the ``Set`` and ``HashSet`` types to ``any``, because it would be either an array or an object, depending on the key type. And yes, Louis is right, you need to change [substituteNewClass](https://github.com/cincheo/jsweet/blob/master/transpiler/src/main/java/org/jsweet/transpiler/extension/RemoveJavaDependenciesAdapter.java#L1182) to create either an array or a map depending on the key type.

Make sure you update the JUnit tests to cover the new cases and keep covering the old ones :)

**Comment by renaudpawlak** [comment]
Almost ready to ship version 2.0.0 RC1... Documentation is already there about how to write transpiler extensions: https://github.com/cincheo/jsweet/blob/master/doc/jsweet-language-specifications.md#extending-the-transpiler.

**Comment by DavidRigglemanININ** [comment]
Awesome. I had some other higher priority stuff in the last month I've had to work on, so I didn't really have much time to work on this. Also, as I was considering the comment about having to map the Set and HashSet types to any, and this made me think that maybe the best option is writing my own StringSet class in Java (backed by a HashMap) that then generates TS so that the code would be typed much stronger then just using the any type. This would be especially true for cases where I might write methods that return a set.
